File: examples/nodejs/_4_ObjectAtomic/publisher.js
const SharedPubSub = require('SharedPubSub');
const sleep = (ms) => new Promise(resolve => setTimeout(resolve, ms));

const publisher = new SharedPubSub.Publisher_ExampleClassAtomic("ObjectAtomic");

async function loop() {
    while (true) {
        /*--------- Example 1 : Publish  ---------*/
        let exampleObject = publisher.readValue();
        exampleObject.setValue1(exampleObject.getValue1()+1);
        exampleObject.setValue2(exampleObject.getValue2()+0.1);
        exampleObject.setValue3(exampleObject.getValue3()+10);
        publisher.publish(exampleObject);
        exampleObject.printValues();
        await sleep(1000)
        /*--------- Example 2 : Get raw pointer, change value and notify  ---------*/
        // Although it's possible, we will not do raw pointer examples
    }
}

loop().catch(console.error);

File: examples/python/_4_ObjectAtomic/publisher.py
from SharedPubSub import *
from time import sleep

publisher = Publisher_ExampleClassAtomic("ObjectAtomic")

while(True):

    ### Example 1 : Publish ###
    exampleObjectAtomic = publisher.readValue()
    exampleObjectAtomic.setValue1(exampleObjectAtomic.getValue1()+1)
    exampleObjectAtomic.setValue2(exampleObjectAtomic.getValue2()+0.1)
    exampleObjectAtomic.setValue3(exampleObjectAtomic.getValue3()+10)
    publisher.publish(exampleObjectAtomic)
    exampleObjectAtomic.printValues()
    sleep(1)

    ### Example 2 : Get raw pointer, change value and notify ###
    rawExampleObjectAtomic = publisher.rawValue()
    rawExampleObjectAtomic.setValue1(rawExampleObjectAtomic.getValue1()+1)
    rawExampleObjectAtomic.setValue2(rawExampleObjectAtomic.getValue2()+0.1)
    rawExampleObjectAtomic.setValue3(rawExampleObjectAtomic.getValue3()+10)
    publisher.notifyAll()
    rawExampleObjectAtomic.printValues()
    sleep(1)
